runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/Endpoint.kt

```diff
@@ -27,9 +23,52 @@ import aws.smithy.kotlin.runtime.http.Url
  * is expected to be mutable and the client cannot modify the endpoint correctly, the operation
  * will likely fail.
  */
-public data class Endpoint(
-    public val uri: Url,
-    public val isHostnameImmutable: Boolean = false,
-) {
+public class Endpoint {
+    /**
+     * The base URL endpoint clients will use to make API calls to e.g. "api.myservice.com".
+     * NOTE: Only `scheme`, `port`, `host` `path`, and `parameters` are valid. Other URL elements are ignored.
+     */
+    public val uri: Url
+
+    /**
+     * Flag indicating whether the hostname can be modified by the SDK client.
+     */
+    public val isHostnameImmutable: Boolean
+
+    /**
+     * A map of additional HTTP headers to be set when making calls against this endpoint.
+     */
+    public val headers: Map<String, List<String>>
+
+    /**
+     * A grab-bag property map of endpoint attributes. The values here are only set when the endpoint returned from
+     * evaluating a ruleset.
+     */
+    @InternalApi
+    public val attributes: Map<String, Any>
```

Comment:
We have a typed property bag [Attributes](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/utils/common/src/aws/smithy/kotlin/runtime/util/Attributes.kt) already. Is there any reason to not leverage that here?

Reply:
Same comment as in `headers`.

Reply:
Yeah in this case is there any reason to not use `Attributes` directly though? It's effectively the same thing except it's a typed property bag rather than untyped. I care less about headers than I do this I think 🤔